sys/include/riotboot/hdr.h

```diff
@@ -52,6 +54,8 @@ typedef struct {
     uint32_t magic_number;      /**< Header magic number (always "RIOT")              */
     uint32_t version;           /**< Integer representing the partition version       */
     uint32_t start_addr;        /**< Address after the allocated space for the header */
+    uint32_t img_size;              /**< Firmware size                                */
+    uint8_t digest[SHA256_DIGEST_LENGTH]; /**< sh256 of image                         */
```

Comment:
The problem is that this will break updates for devices that are deployed in the field with the old riotboot header since we have no header versioning. (or should we use the `magic_number` for that?)

Would it make sense to make those optional by wrapping them in `#ifdef MODULE_RIOTBOOT_SLOT_VERIFY_SHA256`?

Reply:
Hmm, I'd say this is backward compatible since we are just appending fields?

Reply:
`riotboot_slot_validate()` will fail as `chksum` got moved

Reply:
We could do

```C
typedef struct {
    uint32_t magic_number;      /**< Header magic number (always "RIOT")              */
    uint32_t version;           /**< Integer representing the partition version       */
    uint32_t start_addr;        /**< Address after the allocated space for the header */
    uint32_t chksum_legacy;     /**< Checksum of (first 3 words) for compat with old riotboot */
    uint32_t img_size;          /**< Firmware size                                    */
    uint8_t digest[SHA256_DIGEST_LENGTH]; /**< sh256 of image                         */
    uint32_t chksum;            /**< Checksum of riotboot_hdr                         */
} riotboot_hdr_t;

Reply:
?
```C
typedef struct {
    uint32_t magic_number;      /**< Header magic number (always "RIOT")              */
    uint32_t version;           /**< Integer representing the partition version       */
    uint32_t start_addr;        /**< Address after the allocated space for the header */
#if IS_USED(MODULE_RIOTBOOT_DIGEST)
    uint32_t chksum_legacy;     /**< Checksum of first 3 words for compat with old riotboot */
    uint32_t img_size;          /**< Firmware size                                    */
    uint8_t digest[SHA256_DIGEST_LENGTH]; /**< sh256 of image                         */
#endif
    uint32_t chksum;            /**< Checksum of riotboot_hdr                         */
} riotboot_hdr_t;
```

I guess the `genhdr` tools could be modified to output one of those two versions, it would simply be a matter of truncating what is written.

I think what is needed is agreement on whether it's a good idea to include the digest by default or keep it as an optional feature. If going with the first one then we can just avoid the extra `riotboot_digest` module.

Reply:
It could be useful for e.g. #17379 where the old firmware is backed-up to external storage before the update.

Reply:
Offline discussion with @kaspar030 and @bergzand they are in favor ina adding it by default, I'll go with keeping a legacy checksum then